### PR TITLE
updated harvesting scripts

### PR DESCRIPTION
**Pull request by GaelTouquet**
Voici les changements que j'ai fait, dis-moi ce que tu en penses!

### CODE REVIEW

File: H2TauTau/scripts/harvest.py [hunk 1]
```diff
@@ -205,9 +205,6 @@ def get_options():
     parser.add_option("-s", "--subdir-pattern", dest="subdir_pattern",
                       default='*',
                       help='subdir pattern')
-    parser.add_option("-S", "--skim", dest="skim", action='store_true',
```
Comment (lucastorterotot) [inline H2TauTau/scripts/harvest.py:208]: So you choose to remove the skimming feature?

File: H2TauTau/scripts/harvest.py [hunk 2]
```diff
@@ -243,11 +225,8 @@ def harvest(src, subdir_pattern='*', tgz_pattern='*', skim=False, apply_ff=False
         print ds.dest
         ds.unpack()
         ds.hadd()
-        if skim:
-            ogpath = '{}/0000/{}/NtupleProducer/tree.root'.format(ds.dest,ds.dest[:ds.dest.find('/')])
-            newpath = '{}/tree.root'.format(ds.dest[:ds.dest.find('/')])
-            skimntuple(ogpath,newpath)
-            os.system('rm -rf {}'.format(ds.dest))
+        os.system('cp {dirname}/0000/{compname}/NtupleProducer/tree.root {dirname}/tree.root'.format(dirname=ds.dest,compname=ds.dest[ds.dest.find('/'):]))
```
Comment (lucastorterotot) [inline H2TauTau/scripts/harvest.py]: This could work if there is not too many jobs but sometimes you get
```
$ ls $dirname
0000
0001
...
```
so maybe one may re-think these lines.

File: H2TauTau/scripts/multiHarvest.py [hunk 3]
```diff
@@ -7,17 +7,14 @@ def get_options():
     import os
     import sys
     from optparse import OptionParser
-    usage = "usage: %prog [options] <src_dir>"
+    usage = "usage: python multiHarvest.py [options] \n example: python multiHarvest.py -g HiggsVBF125 -l sysCheck0204_nominal -d R -u gtouquet"
```
Comment (lucastorterotot) [inline H2TauTau/scripts/multiHarvest.py]: :+1:

File: H2TauTau/scripts/multiHarvest.py [hunk 4]
```diff
@@ -26,13 +23,13 @@ def get_options():
...
-                      help='Heppy cfg file used to produce the samples.')
-    parser.add_option("-g", "--grep", dest = "cut_on_sample_names",
+                      help='The prod label or part of prod labels to be selected.')
+    parser.add_option("-C", "--component", dest = "cut_on_sample_names",
                       default="''",
                       help='Harvest only samples containing this string.')
     parser.add_option("-d", "--date", dest = "select_date",
                       default='',
-                      help='Harvest only samples submitted on this date. If set to R, look for the most recent job.')
+                      help='Harvest on2ly samples submitted on this date. If set to R, look for the most recent job.')
```
Comment (lucastorterotot) [inline H2TauTau/scripts/multiHarvest.py]: Revert this change.

File: H2TauTau/scripts/multiHarvest.py [hunk 5]
```diff
@@ -26,13 +23,13 @@ def get_options():
                       help='Version of CMSSW used to produce the samples.')
     parser.add_option("-l", "--prod_label", dest = "prod_label",
                       default='diTau_2018_modular_cfg',
-                      help='Heppy cfg file used to produce the samples.')
-    parser.add_option("-g", "--grep", dest = "cut_on_sample_names",
+                      help='The prod label or part of prod labels to be selected.')
+    parser.add_option("-C", "--component", dest = "cut_on_sample_names",
```
Comment (lucastorterotot) [inline H2TauTau/scripts/multiHarvest.py:27]: Ok for the way it is used now. It used to pass `cut_on_sample_names` to the `grep` command on paths so it could also have been used to select more generally directories to harvest.

File: H2TauTau/scripts/multiHarvest.py [hunk 6]
```diff
@@ -53,63 +50,46 @@ def multithreadmap(f,X,ncores=20, **kwargs):
...
+selected_dirs = []
+for prod_label_dir in matching_prod_label_dirs:
+    matching_comp_dirs = find_dirs_in_dir(prod_label_dir, to_match=options.cut_on_sample_names)
+    for comp_dir in matching_comp_dirs:
+        matched_dirs = find_dirs_in_dir(comp_dir)
+        if options.select_date == 'R' :
+            matched_dirs = [matched_dirs[-1]]
+        else :
+            matched_dirs = [dirname for dirname in matching_comp_dirs if options.select_date in line]
+        selected_dirs.extend(matched_dirs)
```
Comment (lucastorterotot) [inline H2TauTau/scripts/multiHarvest.py:74]: Any risk to get a directory twice? If so, use a set instead of a list.

File: H2TauTau/scripts/multiHarvest.py [hunk 7]
```diff
@@ -53,63 +50,46 @@ def multithreadmap(f,X,ncores=20, **kwargs):
...
+
+###formatting
+selected_dirs = [dirname[dirname.find('store')-1:] for dirname in selected_dirs]
```
Comment (lucastorterotot) [inline H2TauTau/scripts/multiHarvest.py:77]: If a set is used this line is still fine.

File: H2TauTau/scripts/multiHarvest.py [hunk 8]
```diff
@@ -53,63 +50,46 @@ def multithreadmap(f,X,ncores=20, **kwargs):
     p.terminate()
     return(Xout)
 
+def find_dirs_in_dir(basepath, to_match=''):
+    if to_match == '':
```
Comment (lucastorterotot) [inline H2TauTau/scripts/multiHarvest.py]: This condition is not necessary if you set `to_match` default value to `"''"`, as is the default value for `options.cut_on_sample_names` as `echo random | grep ''` returns `random`.

**Review by lucastorterotot — CHANGES_REQUESTED** [review]
Some small improvements could be done and some changes are to be reverted.

File: H2TauTau/scripts/harvest.py (hunk 1 above)
Comment (GaelTouquet) [reply H2TauTau/scripts/harvest.py:208]: yes, skimming feature is already implemented at Heppy level. Plus hardcoding skimming here can lead to confusion since the skimming is very likely to change and adapt regularly (as is right now).

File: H2TauTau/scripts/harvest.py (hunk 2 above)
Comment (GaelTouquet) [reply H2TauTau/scripts/harvest.py]: I have not ever experienced such a case, could you share an example? Would you know what are the thresholds/rules that lead to this case?

File: H2TauTau/scripts/multiHarvest.py (hunk 6 above)
Comment (GaelTouquet) [reply H2TauTau/scripts/multiHarvest.py:74]: I don't see how we could get the same directory twice, could you give me an example?

File: H2TauTau/scripts/multiHarvest.py (hunk 5 above)
Comment (GaelTouquet) [reply H2TauTau/scripts/multiHarvest.py:27]: yes but previously the label had to be perfectly right, meaning you could not harvest different systematics of the same production.

File: H2TauTau/scripts/multiHarvest.py (hunk 8 above)
Comment (GaelTouquet) [reply H2TauTau/scripts/multiHarvest.py]: great idea thanks!

File: H2TauTau/scripts/harvest.py (hunk 2 above)
Comment (lucastorterotot) [reply H2TauTau/scripts/harvest.py]: It happens when the amount of crab jobs exceeds 1000.

Comment (GaelTouquet) [reply H2TauTau/scripts/harvest.py]: ok thanks for the information, I will update and push a fix

File: H2TauTau/scripts/multiHarvest.py (hunk 6 above)
Comment (lucastorterotot) [reply H2TauTau/scripts/multiHarvest.py:74]: Ok when looping over a list given by a ls command.

File: H2TauTau/scripts/harvest.py (hunk 2 above)
Comment (GaelTouquet) [reply H2TauTau/scripts/harvest.py]: ok I have pushed a fix, unfortunately I haven't tested it in real conditions as I don't think I have an example of this case. Maybe you could?
Also I don't really have time to work on this anymore, but I think you got the idea of the code design. This latest fix is not very elegant so don't hesitate to submit a fix if you have a better way of coding it.

**Review by lucastorterotot — COMMENTED** [review]
Not sure this is a correct fix, I will look tomorrow or later today.
